qualtran/bloqs/block_encoding/product.py

```diff
@@ -0,0 +1,264 @@
+    r"""Product of a sequence of block encodings.
+
+    Builds the block encoding $B[U_1 * U_2 * \cdots * U_n]$ given block encodings $B[U_1], \ldots, B[U_n]$.
+
+    When each $B[U_i]$ is a $(\alpha_i, a_i, \epsilon_i)$-block encoding of $U_i$, we have that
+    $B[U_1 * \cdots * U_n]$ is a $(\prod_i \alpha_i, n - 1 + \max_i a_i, \sum_i \alpha_i \epsilon_i)$-block
+    encoding of $U_1 * \cdots * U_n$.
+
+    Args:
+        block_encodings: A sequence of block encodings.
+
+    Registers:
+        system: The system register.
+        ancilla: The ancilla register (present only if bitsize > 0).
+        resource: The resource register (present only if bitsize > 0).
+
+    References:
+        [Quantum algorithms: A survey of applications and end-to-end complexities](https://arxiv.org/abs/2310.03011). Dalzell et al. (2023). Ch. 10.2.
+    """
+
+    block_encodings: Tuple[BlockEncoding, ...] = field(
+        converter=lambda x: x if isinstance(x, tuple) else tuple(x), validator=validators.min_len(1)
+    )
+
+    def __attrs_post_init__(self):
+        if not all(u.system_bitsize == self.system_bitsize for u in self.block_encodings):
+            raise ValueError("All block encodings must have the same system size.")
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(
+            system=QAny(self.system_bitsize),
+            ancilla=QAny(self.ancilla_bitsize),  # if ancilla_bitsize is 0, not present
+            resource=QAny(self.resource_bitsize),  # if resource_bitsize is 0, not present
+        )
+
+    @cached_property
+    def system_bitsize(self) -> SymbolicInt:
+        return self.block_encodings[0].system_bitsize
+
+    def pretty_name(self) -> str:
+        return f"B[{'*'.join(u.pretty_name()[2:-1] for u in self.block_encodings)}]"
+
+    @cached_property
+    def alpha(self) -> SymbolicFloat:
+        return prod(u.alpha for u in self.block_encodings)
+
+    @cached_property
+    def ancilla_bitsize(self) -> SymbolicInt:
+        return smax(u.ancilla_bitsize for u in self.block_encodings) + len(self.block_encodings) - 1
+
+    @cached_property
+    def resource_bitsize(self) -> SymbolicInt:
+        return ssum(u.resource_bitsize for u in self.block_encodings)
+
+    @cached_property
+    def epsilon(self) -> SymbolicFloat:
+        return ssum(u.alpha * u.epsilon for u in self.block_encodings)
+
+    @property
+    def target_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("system"),)
+
+    @property
+    def junk_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("resource"),)
+
+    @property
+    def selection_registers(self) -> Tuple[Register, ...]:
+        return (self.signature.get_right("ancilla"),)
+
+    @property
+    def signal_state(self) -> PrepareOracle:
+        """This method will be implemented in the future after PrepareOracle is updated for the BlockEncoding interface."""
```

Comment:
nit: this can be a line comment, and wrap to 100 chars.

If this isn't going to be immediately remedied (although I suspect it might be), it would be good practice to link to an issue as well